[OCPBUGS-35431] Add SNO installation method for LPAR IBM Z (#77425)
openshift/openshift-docs: modules/install-sno-ibm-z-lpar.adoc

// This is included in the following assemblies:
//
// installing_sno/install-sno-installing-sno.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-sno-on-ibm-z-lpar_{context}"]
= Installing {sno} in an LPAR on {ibm-z-title} and {ibm-linuxone-title}

.Prerequisites

> * You have installed `podman`.

Reviewer: Why do we need podman ?
I would remove this line, until it is required.

> * You have an existing `install-config.yaml` file.

Reviewer: Why do we require to have an existing "install-config.yaml" ?
Because, It is described below how to "Prepare the `install-config.yaml` file"
I would remove this line.

ignition.config.url=<rhcos_ign>:8080/ignition/bootstrap-in-place-for-live-iso.ign \// <2>
ip=encbdd0:dhcp::02:00:00:02:34:02 <3>
rd.znet=qeth,0.0.bdd0,0.0.bdd1,0.0.bdd2,layer2=1 \
rd.dasd=0.0.4411 \// <4>
rd.zfcp=0.0.8001,0x50050763040051e3,0x4000406300000000 \// <5>
zfcp.allow_lun_scan=0 \
rd.luks.options=discard
----
<1> For the `coreos.live.rootfs_url=` artifact, specify the matching `rootfs` artifact for the `kernel`and `initramfs` you are booting. Only HTTP and HTTPS protocols are supported.
<2> For the `ignition.config.url=` parameter, specify the Ignition file for the machine role. Only HTTP and HTTPS protocols are supported.
<3> For the `ip=` parameter, assign the IP address automatically using DHCP or manually as described in "Installing a cluster with z/VM on {ibm-z-name} and {ibm-linuxone-name}".
<4> For installations on DASD-type disks, use `rd.dasd=` to specify the DASD where {op-system} is to be installed. Omit this entry for FCP-type disks.
<5> For installations on FCP-type disks, use `rd.zfcp=<adapter>,<wwpn>,<lun>` to specify the FCP disk where {op-system} is to be installed. Omit this entry for DASD-type disks.
+
Leave all other parameters unchanged.

> . Transfer the following artifacts, files, and images to z/VM. For example by using FTP:

Reviewer 2: It should be LPAR, not z/VM

** `kernel` and `initramfs` artifacts
** Parameter files
** {op-system} images
+
For details about how to transfer the files with FTP and boot, see link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/performing_a_standard_rhel_8_installation/installing-in-an-lpar_installing-rhel[Installing in an LPAR].

> . Boot the machine

Reviewer 2: Since this is a process with bootstrap separate, we have to boot both the bootstrap and then subsequently the SNO LPAR (or the master) 

Consider this line here instead. 

```
Boot the bootstrap and master machine with their respective ignition config and boot params.
```

Author: Discussed with @smolin-de that "Boot the machine" is sufficient.

// This is included in the following assemblies:
//
// installing_sno/install-sno-installing-sno.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-sno-on-ibm-z-lpar_{context}"]
= Installing {sno} in an LPAR on {ibm-z-title} and {ibm-linuxone-title}

.Prerequisites

* If you are deploying a single-node cluster there are zero compute nodes; the Ingress Controller pods run on the control plane nodes. In single-node cluster deployments, you must configure your application ingress load balancer to route HTTP and HTTPS traffic to the control plane nodes. See the _Load balancing requirements for user-provisioned infrastructure_ section for more information.
* When you create the Kubernetes manifest files, ensure that the `mastersSchedulable` parameter in the `<installation_directory>/manifests/cluster-scheduler-02-config.yml` file is set to `true`, as shown in the following `spec` stanza:
+
[source,yaml]
----
spec:

>   mastersSchedulable: false

Reviewer 2: Shouldn't this value be shown as true? We are telling the users to set the value to true

Author: True 👍

. Create a parameter file for the control plane in an LPAR:
+
.Example parameter file for the control plane machine
+
[source,terminal]
----
cio_ignore=all,!condev rd.neednet=1 \
console=ttysclp0 \
coreos.inst.install_dev=/dev/<block_device> \
coreos.inst.ignition_url=http://<http_server>/master.ign \// <1>
coreos.live.rootfs_url=http://<http_server>/rhcos-<version>-live-rootfs.<architecture>.img \
ip=<ip>::<gateway>:<netmask>:<hostname>::none nameserver=<dns> \
rd.znet=qeth,0.0.1140,0.0.1141,0.0.1142,layer2=1,portno=0 \
rd.dasd=0.0.4411 \
rd.zfcp=0.0.8001,0x50050763040051e3,0x4000406300000000 \
zfcp.allow_lun_scan=0
----
<1> Specify the location of the `master.ign` config file. Only HTTP and HTTPS protocols are supported.

. Transfer the following artifacts, files, and images to the LPAR. For example by using FTP:

** `kernel` and `initramfs` artifacts
** Parameter files
** {op-system} images
+
For details about how to transfer the files with FTP and boot, see link:https://access.redhat.com/documentation/en-us/red_hat_enterprise_linux/8/html/performing_a_standard_rhel_8_installation/installing-in-an-lpar_installing-rhel[Installing in an LPAR].

> . Boot the machine.

Reviewer: ```suggestion
. Boot the bootstrap machine.
. Boot the control plane machine.
```

// This is included in the following assemblies:
//
// installing_sno/install-sno-installing-sno.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-sno-on-ibm-z-lpar_{context}"]
= Installing {sno} in an LPAR on {ibm-z-title} and {ibm-linuxone-title}

.Prerequisites

> * If you are deploying a single-node cluster there are zero compute nodes; the Ingress Controller pods run on the control plane nodes. In single-node cluster deployments, you must configure your application ingress load balancer to route HTTP and HTTPS traffic to the control plane nodes. See the _Load balancing requirements for user-provisioned infrastructure_ section for more information.

Reviewer 3: ```suggestion
* If you are deploying a single-node cluster there are zero compute nodes, the Ingress Controller pods run on the control plane nodes. In single-node cluster deployments, you must configure your application ingress load balancer to route HTTP and HTTPS traffic to the control plane nodes. See the _Load balancing requirements for user-provisioned infrastructure_ section for more information.
```

.Procedure

. Set the {product-title} version by running the following command:
+
[source,terminal]
----
$ OCP_VERSION=<ocp_version> <1>
----
+
<1> Replace `<ocp_version>` with the current version. For example, `latest-{product-version}`.

. Set the host architecture by running the following command:
+
[source,terminal]
----
$ ARCH=<architecture> <1>
----
<1> Replace `<architecture>` with the target host architecture `s390x`.

> . Download the {product-title} client (`oc`) and make it available for use by entering the following commands:

Reviewer 3: I see that this is following the model of the existing content. It would be good to update all three to use a single command per step, but that is out of scope here. Suggest adding it as a backlog docs enhancement.

Author: Will add this and the next comment to the post GA backlog as this applies to x86 procedures too.

+
[source,terminal]
----
$ curl -k https://mirror.openshift.com/pub/openshift-v4/${ARCH}/clients/ocp/${OCP_VERSION}/openshift-client-linux.tar.gz -o oc.tar.gz
----
+
[source,terminal]
----
$ tar zxvf oc.tar.gz
----
+
[source,terminal]
----
$ chmod +x oc
----

> . Download the {product-title} installer and make it available for use by entering the following commands:

Reviewer 3: Guessing this would need to change in all three, can address later if that's better.

```suggestion
. Download the {product-title} installation program and make it available for use by entering the following commands:
```

** Downloaded {op-system-base} live `kernel`, `initramfs`, and `rootfs` artifacts
** Ignition files

. Create a parameter file for the control plane virtual machine:
+
.Example parameter file for the control plane machine
+
[source,terminal]
----
cio_ignore=all,!condev rd.neednet=1 \
console=ttysclp0 \
coreos.inst.install_dev=/dev/sda \
coreos.inst.ignition_url=http://<http_server>/master.ign \// <1>
coreos.live.rootfs_url=http://<http_server>/rhcos-<version>-live-rootfs.<architecture>.img \// <2>
ip=<ip>::<gateway>:<netmask>:<hostname>::none nameserver=<dns> \// <3>
rd.znet=qeth,0.0.1140,0.0.1141,0.0.1142,layer2=1,portno=0 \
rd.dasd=0.0.4411 \// <4>
rd.zfcp=0.0.8001,0x50050763040051e3,0x4000406300000000 \// <5>
zfcp.allow_lun_scan=0
----
<1> Specify the location of the `master.ign` config file. Only HTTP and HTTPS protocols are supported.
<2> For the `coreos.live.rootfs_url=` artifact, specify the matching `rootfs` artifact for the `kernel`and `initramfs` you are booting. Only HTTP and HTTPS protocols are supported.
<3> For the `ip=` parameter, assign the IP address manually as described in "Installing a cluster in an LPAR on {ibm-z-name} and {ibm-linuxone-name}".

> <4> For installations on DASD-type disks, use `rd.dasd=` to specify the DASD where {op-system} is to be installed. Omit this entry for FCP-type disks.

Reviewer: In this example we boot from FC disk "install_dev"=/dev/sda.
```suggestion
<4> For installations on DASD-type disks, use `rd.dasd=` to specify the DASD where {op-system} is to be installed. Change the 'install_dev' parameter to 'coreos.inst.install_dev=/dev/dasda' and omit this entry for FCP-type disks.
```

// This is included in the following assemblies:
//
// installing_sno/install-sno-installing-sno.adoc

:_mod-docs-content-type: PROCEDURE
[id="installing-sno-on-ibm-z-lpar_{context}"]
= Installing {sno} in an LPAR on {ibm-z-title} and {ibm-linuxone-title}

.Prerequisites

* If you are deploying a single-node cluster there are zero compute nodes; the Ingress Controller pods run on the control plane nodes. In single-node cluster deployments, you must configure your application ingress load balancer to route HTTP and HTTPS traffic to the control plane nodes. See the _Load balancing requirements for user-provisioned infrastructure_ section for more information.

.Procedure

. Set the {product-title} version by running the following command:
+
[source,terminal]
----
$ OCP_VERSION=<ocp_version> <1>
----
+

> <1> Replace `<ocp_version>` with the current version, for example, `latest-{product-version}`

Reviewer 4: Needs punctuation:

<1> Replace `<ocp_version>` with the current version. For example, use `latest-{product-version}`.

Author: These modules have the same sentence I'll add them to the PR
modules/install-sno-ibm-z-kvm.adoc
modules/install-sno-ibm-z.adoc

<1> Add the cluster domain name.
<2> Set the `compute` replicas to `0`. This makes the control plane node schedulable.
<3> Set the `controlPlane` replicas to `1`. In conjunction with the previous `compute` setting, this setting ensures the cluster runs on a single node.
<4> Set the `metadata` name to the cluster name.
<5> Set the `networking` details. OVN-Kubernetes is the only allowed network plugin type for single-node clusters.
<6> Set the `cidr` value to match the subnet of the {sno} cluster.
<7> Copy the {cluster-manager-url-pull} and add the contents to this configuration setting.
<8> Add the public SSH key from the administration host so that you can log in to the cluster after installation.

. Generate {product-title} assets by running the following commands:
+
[source,terminal]
----
$ mkdir ocp
----
+
[source,terminal]
----
$ cp install-config.yaml ocp
----

. Change to the directory that contains the {product-title} installation program and generate the Kubernetes manifests for the cluster:
+
[source,terminal]
----
$ ./openshift-install create manifests --dir <installation_directory> <1>
----
+
<1> For `<installation_directory>`, specify the installation directory that

> contains the `install-config.yaml` file you created.

Reviewer 4: You can remove the line break and include the full sentence on line 123, unless there was a specific reason to add a line :)
